coordinator/src/main/scala/com/socrata/datacoordinator/service/Router.scala

```diff
@@ -0,0 +1,67 @@
+package com.socrata.datacoordinator.service
+
+import com.socrata.datacoordinator.id.{RollupName, DatasetId}
+import com.socrata.datacoordinator.resources.SodaResource
+import com.socrata.http.server._
+import com.socrata.http.server.responses._
+import com.socrata.http.server.routing.{Extractor, SimpleRouteContext}
+import com.socrata.http.server.routing.SimpleRouteContext._
+import com.socrata.datacoordinator.common.util.DatasetIdNormalizer._
+
+case class Router(parseDatasetId: String => Option[DatasetId],
+              notFoundDatasetResource: Option[String] => SodaResource,
+              datasetResource: DatasetId => SodaResource,
+              datasetSchemaResource: DatasetId => SodaResource,
+              datasetRollupResource: DatasetId => SodaResource,
+              secondaryManifestsResource: Option[String] => SodaResource,
+              datasetSecondaryStatusResource: (Option[String], DatasetId) => SodaResource,
+              secondariesOfDatasetResource: DatasetId => SodaResource,
+              versionResource: SodaResource) {
+
+  type OptString = Option[String]
+
+
+
+
+  implicit object DatasetIdExtractor extends Extractor[DatasetId] {
+    def extract(s: String): Option[DatasetId] = parseDatasetId(norm(s))
+  }
+
+  implicit object RollupNameExtractor extends Extractor[RollupName] {
+    def extract(s: String): Option[RollupName] = Some(new RollupName(s))
+  }
+
+  implicit object StringExtractor extends Extractor[OptString]{
+    def extract(s: String): Option[OptString] = if (s.isEmpty) Some(None) else Some(Some(s))
+  }
+
+
+
+  val routes = locally {
+    import SimpleRouteContext._
+    Routes(
+      // "If the thing is parsable as a DatasetId, do something with it, otherwise give a
+      // SODA2 not-found response"
+      Route("/dataset/{OptString}", notFoundDatasetResource),
```

Comment:
This does not appear compatible with the existing functionality, which lists datasets if you hit `/dataset` and returns an error if you hit `/dataset/thisdoesnotexist`.